// Copyright 2022 PingCAP, Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.#pragma once

syntax = "proto3";

package DB.PS.V3.CheckpointProto;

import "common.proto";

/**

A manifest file format V1 consists of the following parts:

- Len|Prefix
- Len|EditsPart
- Len|LocksPart
...

 */

message ManifestFilePrefix {
    uint64 file_format = 1;

    uint64 local_sequence = 11;
    uint64 last_local_sequence = 12;

    // The local timestamp this manifest file is created.
    uint64 create_at_ms = 21;
    WriterInfo writer_info = 22;
}

message ManifestFileEditsPart {
    bool has_more = 1;
    repeated EditRecord edits = 2;
}

message ManifestFileLocksPart {
    bool has_more = 1;
    repeated LockFile locks = 2;
}

enum EditType {
    EDIT_TYPE_UNSPECIFIED = 0;
    EDIT_TYPE_ENTRY = 1;
    EDIT_TYPE_REF = 2;
    EDIT_TYPE_EXTERNAL = 3;
    EDIT_TYPE_DELETE = 4;
}

message EntryDataLocation {
    // There could be heavy duplicates. We rely on compression to reduce file size.
    bytes data_file_id = 1;

    uint64 offset_in_file = 2;
    uint64 size_in_file = 3;
}

message EditRecord {
    EditType type = 1;
    // Note page_id are not valid utf-8 string. Don't use string type.
    bytes page_id = 2;
    bytes ori_page_id = 3;
    uint64 version_sequence = 4;
    uint64 version_epoch = 5;

    EntryDataLocation entry_location = 6;
    uint64 entry_size = 7; // The size of the data (uncompressed)
    uint64 entry_tag = 8;
    uint64 entry_checksum = 9;
    repeated uint64 entry_fields_offset = 10; // The offset to the data (not the offset to the file) for each field.
    repeated uint64 entry_fields_checksum = 11;
}

message LockFile {
    // There could be heavy duplicates. We rely on compression to reduce file size.
    bytes name = 1;
}

